### PR TITLE
Flav/mistral api

### PR DESCRIPTION
**Pull request by flvndvd**
This PR introduces support for the chat block within the Developer Tools to use the newest Mistral models (`mistral-tiny`, `mistral-small`, `mistral-medium`) via their own API.

**Key points to note:**
- Given that the Mistral API is currently in beta, a streamlined yet effective error handling approach has been implemented.
- Presently, this enhancement is confined to the `chat` block.

**Follow-up PRs will come pretty soon to:**
- Support the `llm` block within the Developer Tools.
- Support Mistral embeddings functionality via `mistral-embed` model.
- Necessary adjustments to facilitate Mistral AI integration in the assistants.
- Refinements to ensure Rust types are more generics, to be tailored to each provider.
- Add meta-prompt with user.

### CODE REVIEW

File: core/src/providers/provider.rs [hunk 1]
```diff
@@ -48,9 +51,10 @@ impl FromStr for ProviderID {
             "ai21" => Ok(ProviderID::AI21),
             "azure_openai" => Ok(ProviderID::AzureOpenAI),
             "anthropic" => Ok(ProviderID::Anthropic),
+            "mistralai" => Ok(ProviderID::MistralAI),
```
Comment (spolu) [inline core/src/providers/provider.rs:54]: This string is use visible and not great great. I would go with `mistral` or `mistral_ai` ?

File: front/components/assistant_builder/AssistantBuilder.tsx [hunk 2]
```diff
@@ -80,6 +81,7 @@ const usedModelConfigs = [
   CLAUDE_DEFAULT_MODEL_CONFIG,
   CLAUDE_INSTANT_DEFAULT_MODEL_CONFIG,
   MISTRAL_7B_DEFAULT_MODEL_CONFIG,
+  MISTRAL_SMALL_MODEL_CONFIG,
```
Comment (spolu) [inline front/components/assistant_builder/AssistantBuilder.tsx:84]: This is unrelated to the addition of mistral to core + dev tools let's keep it for the next PR :+1:

File: types/src/front/lib/assistant.ts [hunk 3]
```diff
@@ -78,6 +79,15 @@ export const MISTRAL_7B_DEFAULT_MODEL_CONFIG = {
   largeModel: false,
 } as const;
 
+export const MISTRAL_SMALL_MODEL_CONFIG = {
```
Comment (spolu) [inline types/src/front/lib/assistant.ts:82]: This is unrelated to the addition of mistral to core + dev tools let's keep it for the next PR :+1:

File: types/src/front/lib/api/credentials.ts [hunk 4]
```diff
@@ -55,10 +59,11 @@ export const credentialsFromProviders = (
 
 export const dustManagedCredentials = (): CredentialsType => {
   return {
-    OPENAI_API_KEY: DUST_MANAGED_OPENAI_API_KEY,
     ANTHROPIC_API_KEY: DUST_MANAGED_ANTHROPIC_API_KEY,
-    TEXTSYNTH_API_KEY: DUST_MANAGED_TEXTSYNTH_API_KEY,
     AZURE_OPENAI_API_KEY: DUST_MANAGED_AZURE_OPENAI_API_KEY,
     AZURE_OPENAI_ENDPOINT: DUST_MANAGED_AZURE_OPENAI_ENDPOINT,
+    MISTRAL_AI_API_KEY: DUST_MANAGED_MISTRAL_AI_API_KEY,
```
Comment (spolu) [inline types/src/front/lib/api/credentials.ts:65]: This as well is for the release of Mistral as an assistant and can be kept in the next PR :+1:

File: front/pages/api/w/[wId]/providers/[pId]/models.ts [hunk 5]
```diff
@@ -219,6 +219,46 @@ async function handler(
           res.status(200).json({ models: anthropic_models });
           return;
 
+        case "mistralai":
```
Comment (spolu) [inline front/pages/api/w/[wId]/providers/[pId]/models.ts:222]: Let's def go for `mistral_ai`

File: front/lib/providers.ts [hunk 6]
```diff
@@ -54,6 +54,14 @@ export const modelProviders: ModelProvider[] = [
     chat: true,
     embed: false,
   },
+  {
+    providerId: "mistralai",
```
Comment (spolu) [inline front/lib/providers.ts:58]: `mistral_ai` :+1:

File: core/src/providers/mistralai.rs [hunk 7]
```diff
@@ -0,0 +1,680 @@
...
+
+use super::llm::ChatFunction;
+use super::tiktoken::tiktoken::{decode_async, encode_async, tokenize_async};
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+#[serde(rename_all = "lowercase")]
+pub enum MistralAIChatMessageRole {
+    Assistant,
+    System,
+    User,
```
Comment (spolu) [inline core/src/providers/mistralai.rs:33]: No User right?

File: core/src/providers/mistralai.rs [hunk 8]
```diff
@@ -0,0 +1,680 @@
...
+    pub _type: Option<String>,
+    pub code: Option<String>,
+    pub message: String,
+    pub object: String,
+    pub param: Option<String>,
+}
+
+impl Error {
+    pub fn message(&self) -> String {
+        format!("MistralAIAPIError: [{:?}] {}", self._type, self.message,)
```
Comment (spolu) [inline core/src/providers/mistralai.rs:108]: We can drop the API here. We should remove it from OpenAI as well I think

File: core/src/providers/mistralai.rs [hunk 9]
```diff
@@ -0,0 +1,680 @@
...
+        MistralAILLM { id, api_key: None }
+    }
+
+    fn chat_uri(&self) -> Result<Uri> {
+        Ok(format!("https://api.mistral.ai/v1/chat/completions",).parse::<Uri>()?)
+    }
+
+    fn tokenizer(&self) -> Arc<Mutex<CoreBPE>> {
+        // TODO(flav): Pending further insights into Mistral tokenizer behavior.
+        return cl100k_base_singleton();
```
Comment (spolu) [inline core/src/providers/mistralai.rs:142]: Can you just check the size of the vocab of Mistral. Maybe we want to use the smaller tokenizer from tiktoken vs 100k to have a better estimate

File: core/src/providers/mistralai.rs [hunk 10]
```diff
@@ -0,0 +1,680 @@
...
+        };
+        builder = match builder.header("Content-Type", "application/json") {
+            Ok(b) => b,
+            Err(_) => return Err(anyhow!("Error creating streamed client to Mistral AI")),
+        };
+
+        // TODO(flav): Handle `safe_mode`.
+
+        let mut body = json!({
+            "messages": messages,
```
Comment (spolu) [inline core/src/providers/mistralai.rs:177]: I presume this will break if we pass a user message. Let's add the prompting layer ala Anthrpic :+1:

**Review by spolu — COMMENTED** [review]
Loooks great. Left a few comments. Should be easy to fix.

File: core/src/providers/mistralai.rs (hunk 10 above)
Comment (flvndvd) [reply core/src/providers/mistralai.rs:226]: Addressed in `to_mistral_messages` function.

File: core/src/providers/mistralai.rs (hunk 9 above)
Comment (flvndvd) [reply core/src/providers/mistralai.rs:191]: I posted a message on the Slack channel asking about the vocabulary size used in the new Mistral models. I will provide an update here once I receive a response with that information.

File: core/src/providers/mistralai.rs [hunk 11]
```diff
@@ -0,0 +1,731 @@
...
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+struct ChatMessage {
+    pub content: Option<String>,
+    pub role: MistralAIChatMessageRole,
+}
+
+impl TryFrom<BaseChatMessage> for ChatMessage {
+    type Error = anyhow::Error;
+
+    fn try_from(value: BaseChatMessage) -> Result<Self, Self::Error> {
```
Comment (spolu) [inline core/src/providers/mistralai.rs:79]: I would take a reference here and clone the value.content in that function

File: core/src/providers/mistralai.rs (hunk 9 above)
Comment (spolu) [reply core/src/providers/mistralai.rs:191]: :+1:

File: core/src/providers/mistralai.rs [hunk 12]
```diff
@@ -0,0 +1,731 @@
...
+    fn chat_uri(&self) -> Result<Uri> {
+        Ok(format!("https://api.mistral.ai/v1/chat/completions",).parse::<Uri>()?)
+    }
+
+    fn to_mistral_messages(
+        &self,
+        messages: &Vec<BaseChatMessage>,
+    ) -> Result<Vec<ChatMessage>, anyhow::Error> {
+        let mistral_messages: Result<Vec<ChatMessage>, _> = messages
+            .iter()
```
Comment (spolu) [inline core/src/providers/mistralai.rs:179]: No need to clone if the try_from takes a reference?

File: front/lib/providers.ts [hunk 13]
```diff
@@ -54,6 +54,14 @@ export const modelProviders: ModelProvider[] = [
     chat: true,
     embed: false,
   },
+  {
+    providerId: "mistral_ai",
```
Comment (spolu) [inline front/lib/providers.ts:58]: I'm wondering if we want `mistral_ai` or just `mistral` the `_ai` does not seem to serve any purpose given how unambiguous "mistral" is?

**Review by spolu — APPROVED** [review]

File: front/lib/providers.ts (hunk 13 above)
Comment (flvndvd) [reply front/lib/providers.ts]: Updated all occurrences to go with `mistral`.

File: core/src/providers/mistralai.rs (hunk 9 above)
Comment (flvndvd) [reply core/src/providers/mistralai.rs:188]: Addressed in https://github.com/dust-tt/dust/pull/2920
